---
title: "Microbiome- DADA2 processing"
author: "Omri Nahor"
date: "2025-10-01"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

---

```{r}
if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
if (!requireNamespace("dada2", quietly = TRUE)) BiocManager::install("dada2")
if (!requireNamespace("phyloseq", quietly = TRUE)) BiocManager::install("phyloseq")

library(dada2)
library(phyloseq)
```

```{r}
setwd("")
path <- ""
list.files(path)
```

```{r}
fnFs <- sort(list.files(path, pattern="_L001_R1_001.fastq.gz", full.names = TRUE, recursive = TRUE))
fnRs <- sort(list.files(path, pattern="_L001_R2_001.fastq.gz", full.names = TRUE, recursive = TRUE))
sample.names <- sapply(strsplit(basename(fnFs), "_L001_R1_001.fastq.gz"), `[`, 1)
```

```{r}
plotQualityProfile(fnFs[1:2])
plotQualityProfile(fnRs[1:2])
```

```{r}
filtFs <- file.path(path, "filtered", paste0(sample.names, "_F_filt.fastq.gz"))
filtRs <- file.path(path, "filtered", paste0(sample.names, "_R_filt.fastq.gz"))

out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen = c(250, 225),
                     maxN = 0, maxEE = c(2, 2), truncQ = 2, trimLeft = c(19, 20),
                     compress = TRUE, multithread = FALSE)
head(out)
```

```{r}
errF <- learnErrors(filtFs, multithread = TRUE)
errR <- learnErrors(filtRs, multithread = TRUE)

plotErrors(errF, nominalQ = TRUE)
plotErrors(errR, nominalQ = TRUE)
```

```{r}
derepFs <- derepFastq(filtFs, verbose = TRUE)
derepRs <- derepFastq(filtRs, verbose = TRUE)
names(derepFs) <- sample.names
names(derepRs) <- sample.names
```

```{r}
dadaFs <- dada(derepFs, err = errF, multithread = TRUE)
dadaRs <- dada(derepRs, err = errR, multithread = TRUE)
```

```{r}
mergers <- mergePairs(dadaFs, derepFs, dadaRs, derepRs, verbose = TRUE, minOverlap = 20)
seqtab <- makeSequenceTable(mergers)
table(nchar(getSequences(seqtab)))
dim(seqtab)
```

```{r}
saveRDS(seqtab, "seqtab.rds")
write.csv(seqtab, "seqtab.csv")
```

```{r}
seqtab.nochim <- removeBimeraDenovo(seqtab, method = "consensus", multithread = TRUE, verbose = TRUE)
dim(seqtab.nochim)
table(nchar(getSequences(seqtab.nochim)))
saveRDS(seqtab.nochim, "seqtab_nochim.rds")
```

```{r}
getN <- function(x) sum(getUniques(x))
track <- cbind(out,
               sapply(dadaFs, getN),
               sapply(dadaRs, getN),
               sapply(mergers, getN),
               rowSums(seqtab.nochim))
colnames(track) <- c("input", "filtered", "denoisedF", "denoisedR", "merged", "nonchim")
rownames(track) <- sample.names
write.csv(track, "track_summary.csv")
```

```{r}
taxa <- assignTaxonomy(seqtab.nochim, "silva_nr99_v138.1_train_set.fa.gz", multithread = TRUE, minBoot = 80)
taxa <- addSpecies(taxa, "silva_species_assignment_v138.1.fa.gz")
saveRDS(taxa, "taxa_assignment.rds")
```

```{r}
ps <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows = FALSE), tax_table(taxa))
dataps <- cbind(as.data.frame(t(ps@otu_table)), as.data.frame(ps@tax_table))
write.csv(dataps, "phyloseq_taxa_table.csv")
```
